OSGeo/gdal
# cmake: Changes to Allow vsihdfs to Build on Ubuntu

## What does this PR do?

This PR add some changes to allow the `vsihdfs` virtual filesystem to build on Ubuntu.  On a freshly-created Ubuntu VM (Lubuntu 21.10), I encountered problems related to the fact that `libhdfs` depends on `libjvm`, so when linking against the former it is required to link against the latter (at least on Ubuntu).

I have made some changes to `findHDFS.cmake` to (try to) force it to look for `libjvm` and link against it.  These changes succeeded for me, although I understand that they might not be appropriate for other systems (I am guessing, but it seems as though MacOS does not behave the same way).  This PR should be consider **work in progress** (and/or closable without offence) because of the fact (1) that it was not test on Windows or MacOS and (2) the fact that the idea itself may not have been correctly implemented (`cmake` is fairly new to me).

I am also happy to keep working on it in response to review comments if there is interest.

## What are related issues/pull requests?

See also: https://github.com/OSGeo/gdal/pull/5112

## Tasklist

 - [ ] Add test case(s)
 - [ ] Adjust for comments
 - [ ] All CI builds and checks have passed

## Environment

Tested in the following environment

* OS:  Lubuntu 21.10
* Compiler: gcc 11.2.0
* cmake: 3.18.4
* openjdk 18-ea 2022-03-15

Testing steps taken:

```bash
export JAVA_HOME=/usr/lib/jvm/java-18-openjdk-amd64/
export LD_LIBRARY_PATH="/usr/lib/jvm/java-18-openjdk-amd64/lib/server/"
mkdir build ; cd build
cmake -DCMAKE_BUILD_TYPE=Release -DHDFS_INCLUDE_DIR=$HOME/local/hadoop-3.3.1/include \
 -DHDFS_LIBRARY=$HOME/local/hadoop-3.3.1/lib/native/libhdfs.so -DGDAL_USE_HDFS=ON \
 -DCMAKE_INSTALL_PREFIX=$HOME/local/gdal -DCMAKE_LIBRARY_PATH="/usr/lib/jvm/java-18-openjdk-amd64/lib/server/" \
 -DBUILD_PYTHON_BINDINGS=OFF \
 ..
make
make install
```
then
```bash
export JAVA_HOME=/usr/lib/jvm/java-18-openjdk-amd64/
export CLASSPATH=$($HOME/Desktop/hadoop-3.3.1/bin/hadoop classpath --glob)
...
~/local/gdal/bin/gdalinfo /vsihdfs/file:/tmp/sample.tif
```
and the last command succeeded.

I should also note that I had to disable the python bindings in order for this to succeed, but that might be unrelated.
